src/transforms/kubernetes/mod.rs

```diff
@@ -1 +1,552 @@
+/// scenario when a pod is deleted every second, we would have ~200kB of probably
+/// not usefull data. Which is acceptable. The benefit of such a long delay
+/// is that we will certainly* have the metadata while we are processing the
+/// remaining logs from the deleted pod.
+const DELETE_DELAY: Duration = Duration::from_secs(10 * 60);
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct KubePodMetadata {
+    #[serde(default = "default_fields")]
+    fields: Vec<String>,
+}
+
+inventory::submit! {
+    TransformDescription::new_without_default::<KubePodMetadata>("kubernetes_pod_metadata")
+}
+
+#[typetag::serde(name = "kubernetes_pod_metadata")]
+impl TransformConfig for KubePodMetadata {
+    fn build(&self, cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        // Main idea is to have a background task which will premptively
+        // acquire metadata for all pods on this node, and with it maintaine
+        // a map of extracted metadata.
+
+        // Construct WatchClient
+        let node = std::env::var(NODE_NAME_ENV)
+            .map_err(|_| BuildError::MissingNodeName { env: NODE_NAME_ENV })?;
+        let wc_config = ClientConfig::in_cluster(node, cx.resolver()).context(WatchClientBuild)?;
+        let watch_client = wc_config.build().context(WatchClientBuild)?;
+
+        // Construct MetadataClient
+        let (reader, writer) = evmap::new();
+        let metadata_client = MetadataClient::new(self, writer, watch_client)?;
+
+        // Run background task
+        cx.executor().spawn_std(async move {
+            let error = metadata_client.run().await;
+            error!(message = "Stopped updating Pod metadata.", reason = %error);
+        });
+
+        // Construct transform
+        Ok(Box::new(KubernetesPodMetadata { metadata: reader }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kubernetes_pod_metadata"
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Failed building WatchClient: {}", source))]
+    WatchClientBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display("Failed building watch stream: {}", source))]
+    WatchStreamBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display(
+        "Missing environment variable {:?} containing node name `spec.nodeName`.",
+        env
+    ))]
+    MissingNodeName { env: &'static str },
+    #[snafu(display("Unknown metadata fields: {:?}", fields))]
+    UnknownFields { fields: Vec<String> },
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
```

Comment:
why does this need to be boxed?

Reply:
Because all `Fn` closures in there are different so we need to turn them into trait objects.

Reply:
Fair, though I do find the usage of closures quite hard to follow. I wonder if there might be a better solution to extract a value from a struct like so.

Reply:
One big `enum` with a variant for every supported field is also a solution. It should be clearer.